[opaque_obj] Allow member accesses on reference types

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172099
* #172092
* #171484
* __->__ #171483

Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

angelayi committed:
[opaque_obj] Add method support

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify a dictionary mapping member name to `MemberType` when registering the opaque object. For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    members={
        "seed": MemberType.GUARDED,
        "get_seed": MemberType.CONSTANT,
        "noisy_inject": MemberType.INLINED,
    },
)
```

There are three `MemberType`s:
- MemberType.GUARDED: Tells torch.compile to guards (via EQUALS_MATCH) on following attributes. We will trigger a recompilation if the attribute changes
- MemberType.INLINED: Tells torch.compile to inline/trace through the method
- MemberType.CONSTANT: Tells torch.compile to inline/trace through the method, but the result is guaranteed to be a constant. If the results change, it should've been guarded on by a field in MemberType.GUARDED. This is introduced when I ran into the following error when inlining `x.get_group(0)`: 
```
Dynamo does not know how to trace the builtin `torch._C._distributed_c10d.pybind11_detail_function_record_v1_system_libstdcpp_gxx_abi_1xxx_use_cxx11_abi_1._resolve_process_group.` This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind).
```

[ghstack-poisoned]

dolpm commented on torch/_library/opaque_object.py:
```diff
@@ -172,7 +218,10 @@ def is_opaque_type(cls: Any) -> bool:
+    except TypeError:
         return False
 
     return torch._C._is_opaque_type_registered(_OPAQUE_TYPES[cls].class_name)
```
this seems more reasonable: 

```py
try:
    return torch._C._is_opaque_type_registered(_OPAQUE_TYPES[cls].class_name)
except TypeError:
    return False
```

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify a dictionary mapping member name to `MemberType` when registering the opaque object. For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    members={
        "seed": MemberType.GUARDED,
        "get_seed": MemberType.CONSTANT,
        "noisy_inject": MemberType.INLINED,
    },
)
```

There are three `MemberType`s:
- MemberType.GUARDED: Tells torch.compile to guards (via EQUALS_MATCH) on following attributes. We will trigger a recompilation if the attribute changes
- MemberType.INLINED: Tells torch.compile to inline/trace through the method
- MemberType.CONSTANT: Tells torch.compile to inline/trace through the method, but the result is guaranteed to be a constant. If the results change, it should've been guarded on by a field in MemberType.GUARDED. This is introduced when I ran into the following error when inlining `x.get_group(0)`: 
```
Dynamo does not know how to trace the builtin `torch._C._distributed_c10d.pybind11_detail_function_record_v1_system_libstdcpp_gxx_abi_1xxx_use_cxx11_abi_1._resolve_process_group.` This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind).
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi commented on torch/_library/opaque_object.py:
```diff
@@ -172,7 +218,10 @@ def is_opaque_type(cls: Any) -> bool:
+    try:
+        if cls not in _OPAQUE_TYPES:
+            return False
+    except TypeError:
```
I ran into an issue where the cls being passed in here is not hashable :// so it resulted in the TypeError

angelayi replied:
hm ok I cant repro the error anymore so i guess ill just remove this

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify a dictionary mapping member name to `MemberType` when registering the opaque object. For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    members={
        "seed": MemberType.GUARDED,
        "get_seed": MemberType.CONSTANT,
        "noisy_inject": MemberType.INLINED,
    },
)
```

There are three `MemberType`s:
- MemberType.GUARDED: Tells torch.compile to guards (via EQUALS_MATCH) on following attributes. We will trigger a recompilation if the attribute changes
- MemberType.INLINED: Tells torch.compile to inline/trace through the method
- MemberType.CONSTANT: Tells torch.compile to inline/trace through the method, but the result is guaranteed to be a constant. If the results change, it should've been guarded on by a field in MemberType.GUARDED. This is introduced when I ran into the following error when inlining `x.get_group(0)`: 
```
Dynamo does not know how to trace the builtin `torch._C._distributed_c10d.pybind11_detail_function_record_v1_system_libstdcpp_gxx_abi_1xxx_use_cxx11_abi_1._resolve_process_group.` This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind).
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -2204,9 +2204,12 @@ def add_torchbind_input(name, value):
+            except (TypeError, AttributeError, pickle.PicklingError) as e:
+                output.writeline(
+                    f'raise TypeError("Failed to pickle opaque type {type(value)} for variable {name}: {str(e)}")'
+                )
```
This is part of the inductor generated code's benchmarking script ([example paste](https://www.internalfb.com/phabricator/paste/view/P2113548477?lines=124)). So whoever is actually running the inductor generated code directly would run into the error. I verified running the file results in the error, but I'm not sure how to test this

zou3519 replied:
oh I see. Let's leave it be, thanks for clarifying

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Allow member accesses on reference types"


Currently we do not allow user code to access the contents of reference-type opaque objects as these are treated like black boxes, and we don't know how to guard on the contents of the black box. This includes attribute accesses and method calls. However in some cases, users do know what fields to properly guard on the reference type, and know which methods do not have side effects. So, we should allow these users to mark these fields as being allowed. 

Specifically this use-case appeared when trying to figure out how to make DeviceMesh an opaque object, where:
* It cannot be a value-type as initializing a DeviceMesh is expensive due to needing to create ProcessGroups
* It cannot be written like a pytree-type as DeviceMesh is not written like a container (it has a lot of internal metadata, like a cache)
* But as a reference type:
  * DeviceMeshes are never created within a program, always created outside of a program and is constant within a run.
  * DeviceMeshes need to be guarded upon to make sure torch.compile recompiles if there is a new DeviceMesh provided
  * Ops rely on method calls like get_rank, get_local_rank, _flatten, _unflatten
  * These methods all return constants – flatten/unflatten returns a DeviceMesh but these can all be created ahead of time and cached. So the results can all be lifted as inputs

This PR allows for users to access attributes/methods on reference-type opaque objects. To do so, the user has to specify 
1. A `guard_fn` which returns a list of values on the object to guard on.
1. A dictionary mapping member name to `MemberType` when registering the opaque object. There are two `MemberType`s:
    a. `MemberType.INLINED`: Tells torch.compile to inline/trace through the method's python source code
    b. `MemberType.USE_REAL`: Tells torch.compile to evaluate this method/attribute based on the real object that is being used for tracing. This can be used for attributes, or for methods which cannot be inlined (maybe due to cpp code) and are guaranteed to always return a constant result.
 
 
For example:
```python
class RNGState:
    def __init__(self, seed):
        self.seed = seed
        self.rng = random.Random(self.seed)

    def get_seed(self):
        return self.seed

    def noisy_inject(self, x):
        return torch.ops._TestOpaqueObject.noisy_inject(x, self)

register_opaque_type(
    RNGState,
    typ="reference",
    guard_fn=lambda obj: [obj.seed]
    members={
        "seed": MemberType.USE_REAL,
        "get_seed": MemberType.USE_REAL,
        "noisy_inject": MemberType.INLINED,
    },
)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela

[ghstack-poisoned]
